File: 54.py
# Spiral Matrix
# Tested but chatGPT but not reviewed by me
matrix1 = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
matrix2 = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]



def extract_spiral(matrix):
    top = 0
    bottom = len(matrix) - 1
    left = 0
    right = len(matrix[0]) - 1
    result = []

    while top <= bottom and left <= right:
        # Traverse top row
        for i in range(left, right + 1):
            result.append(matrix[top][i])
        top += 1

        # Traverse rightmost column
        for i in range(top, bottom + 1):
            result.append(matrix[i][right])
        right -= 1

        # Traverse bottom row
        if top <= bottom:
            for i in range(right, left - 1, -1):
                result.append(matrix[bottom][i])
            bottom -= 1

        # Traverse leftmost column
        if left <= right:
            for i in range(bottom, top - 1, -1):
                result.append(matrix[i][left])
            left += 1

    return result

            

print(extract_spiral(matrix2))

File: 54.js
// 54: Spiral Matrix
// Tested successfully


let matrix1 = [[1,2,3],[4,5,6],[7,8,9]];
let matrix2 = [[1,2,3,4],[5,6,7,8],[9,10,11,12]];
let matrix3 = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]]

function extractBoundary(matrix, boundary_top, boundary_bot, boundary_left, boundary_right)
{
    let res = [];
    for (let j = boundary_left; j <= boundary_right; j++)
    {
        res.push(matrix[boundary_top][j])
    }

    for(let i = boundary_top + 1; i < boundary_bot; i++)
    {
        res.push(matrix[i][boundary_right]);
    }

    if (boundary_bot != boundary_top)
    {
        for(let j = boundary_right; j >= boundary_left; j--)
        {
            res.push(matrix[boundary_bot][j]);
        }
    }


    for(let i = boundary_bot-1; i > boundary_top; i--)
    {
        res.push(matrix[i][boundary_left]);
    }

    return res;
}

function spiralMatrix(matrix)
{

    let numRows = matrix.length;
    let numCols = matrix[0].length;
    let boundary_top = 0;
    let boundary_bot = matrix.length - 1;
    let boundary_left = 0;
    let boundary_right = matrix[0].length-1;
    let result = [];
    while(boundary_top<boundary_bot)
    {
        result.push(...extractBoundary(matrix, boundary_top, boundary_bot, boundary_left, boundary_right));
        boundary_bot--;
        boundary_top++;
        boundary_left++;
        boundary_right--;

    }
    result.push(...extractBoundary(matrix, boundary_top, boundary_bot, boundary_left, boundary_right));
    


    return result;

}

// console.log(extractBoundary(matrix3, 1, 2, 1, 2))
console.log(spiralMatrix(matrix3));